docs/Outputs.md

```diff
@@ -0,0 +1,184 @@
+ 
+## Endpoints, Encoders, and Transports
+
+There are three key components for an external service:
+ - an _endpoint_: it is the component that takes a table and pushes it to the external service.
+ - an _encoder_: it is the component that translates tables into something understandable by the external service.
+ - the _transport_: it is the communication protocol used to communicate with the external service.
+
+In a nutshell, endpoints use encoders to encode data and a transport to push data to external services.
+
+Encoders and transports are written in Go as `values.Function`s.
+For example there will be `json.encode`, as well as `csv.encode` for encoders, and `http.post` and `tcp.send` for transports.
+
+The minimal set that must be implemented is `http.post` and `json.encode`.
+
+Endpoints are implemented in pure Flux.
+
+## Flux
+
+Encoders are provided in Flux as `encode` functions.
+For example the JSON encoder is part of the `json` package.
+Encoders take an object `data` and return its encoding as a string:
+
+```flux
+encode: (data:object) -> string
+```
+
+Communication protocols are also implemented (in Go), so that one can:
+
+```flux
+import "http"
+import "tcp"
+
+import "json"
+import "csv"
+
+data = {a: 1, b: 2}
+
+json_enc = json.encode(data)
+csv_enc = csv.encode(data)
+
+http.post(header: {"Content Type": "application-json"}, url: "http://some/url/", data: json_enc)
+
+conn = tcp.dial(url: "tcp://some/url/")
+tcp.send(connection: conn, data: csv_enc)
+tcp.close(connection: conn)
+```
+
+External services in Flux provide an `endpoint` function for building an endpoint.
+All `endpoint` functions return an object for the endpoint.
+Every `endpoint` function has at least these parameters:
+ - `name` for the user-defined name of the endpoint.
+ - `async` for sending results asynchronously.
+  When working in asynchronous mode, the endpoint will not wait for the answer of the external service to consider the data pushed.
+  The user could choose to set `async: false` when it is of paramount importance that processing advances only when data reached the external services. 
+ - `fail` if the user wants the script to fail when the `Send` method returns an error.
+  If `fail` is set to `false`, the error will only be logged.
+ 
+The endpoint object has at least two fields:
+ - `name`: the endpoint name as specified by the user;
+ - `to`: a function that accepts a stream of tables as input and sends data to the external service.
+
+`to` transformations for endpoints can be implemented using the `map` transformation.
+For example, this could be an implementation for a Slack endpoint:
+
+```flux
+package slack
+
+import "http"
+import "json"
+
+builtin endpoint
+
+// Util functions to create the header and the url from the endpoint configuration.
+header = (token, ...) => ...
+url = (...) => ...
+
+to = (header, url) => (tables=<-) => tables
```

Comment:
I think we should separate the concept of tables from the concept of message creation/sending as much as possible.  

Kapacitor has had so many feature requests/EAR's because it tries to have built-in message packing for all of these services.   

unless i'm misunderstanding the requirements, we are not trying to send bulk data to these services, but rather a single message to the service.  We should provide helpers to create such messages.  

for example, the slack POST API:   https://api.slack.com/methods/chat.postMessage

we should have a helper that generates the JSON perfectly for the given parameters.  

how does a table fit into a slack message?  I think that's less well-defined and up to the user.    Separating these tasks will help us with maintainability and re-usability.  Packaging up a large table into a compact CSV-string version, for example, can be used by probably most alert formats.

Reply:
@aanthony1243 I see you point! I can expose the method used in `map` to the user. I'll do that right away.